subsys/suit/stream/stream_sinks/src/suit_sdfw_recovery_sink.c

```diff
@@ -0,0 +1,366 @@
+	/* Since mram_erase returns void, there is a need for extra check and returning error code
+	 * to handle such case.
+	 */
+	if (NRF_SICR->UROT.UPDATE.STATUS == SICR_UROT_UPDATE_STATUS_CODE_None &&
+	    NRF_SICR->UROT.UPDATE.OPERATION == SICR_UROT_UPDATE_OPERATION_OPCODE_Nop) {
+		return SUIT_PLAT_SUCCESS;
+	} else {
+		return SUIT_PLAT_ERR_IO;
+	}
+}
+
+static suit_plat_err_t schedule_sdfw_recovery_update(const uint8_t *buf, size_t size)
+{
+	int err = 0;
+
+	const struct sdfw_update_blob update_blob = {
+		.manifest_addr =
+			(uintptr_t)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_SIGNED_MANIFEST_OFFSET),
+		.pubkey_addr =
+			(uintptr_t)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_PUBLIC_KEY_OFFSET),
+		.signature_addr =
+			(uintptr_t)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_SIGNATURE_OFFSET),
+		.firmware_addr =
+			(uintptr_t)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET),
+		.max_size = CONFIG_SUIT_SDFW_RECOVERY_UPDATE_MAX_SIZE,
+	};
+
+	LOG_DBG("update_candidate: 0x%08lX", (uintptr_t)buf);
+	LOG_DBG("signed_manifest: 0x%08lX", update_blob.manifest_addr);
+	LOG_DBG("public_key: 0x%08lX", update_blob.pubkey_addr);
+	LOG_DBG("signature: 0x%08lX", update_blob.signature_addr);
+	LOG_DBG("firmware: 0x%08lX", update_blob.firmware_addr);
+	LOG_DBG("max update size: 0x%08X", update_blob.max_size);
+	LOG_DBG("firmware size: 0x%08X (%d)", size, size);
+
+	/* NOTE: SecROM does not use the actual size of new SDFW during update process.
+	 *       However, if SDFW is flashed for the first time, the value is used to limit the
+	 * future maximum SDFW size. Hence, use the maximum allowed size for safety.
+	 */
+	err = sdfw_recovery_update(&update_blob);
+
+	if (err) {
+		LOG_ERR("Failed to schedule SDFW Recovery update: %d", err);
+		return SUIT_PLAT_ERR_CRASH;
+	}
+
+	LOG_INF("SDFW Recovery update scheduled");
+
+	return SUIT_PLAT_SUCCESS;
+}
+
+static sdf_sink_err_t check_update_candidate(const uint8_t *buf, size_t size)
+{
+	uint8_t *candidate_binary_start =
+		(uint8_t *)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET);
+	uint8_t *candidate_digest_in_manifest =
+		(uint8_t *)(buf + CONFIG_SUIT_SDFW_RECOVERY_UPDATE_DIGEST_OFFSET);
+	uint8_t *current_sdfw_recovery_digest =
+		(uint8_t *)(NRF_SICR->UROT.RECOVERY.SM.TBS.FW.DIGEST);
+
+	/* First check if calculated digest of candidate matches the digest from Signed Manifest */
+	digest_sink_err_t err =
+		verify_digest(candidate_binary_start,
+			      size - (size_t)CONFIG_SUIT_SDFW_RECOVERY_UPDATE_FIRMWARE_OFFSET,
+			      PSA_ALG_SHA_512, candidate_digest_in_manifest);
+
+	if (err != SUIT_PLAT_SUCCESS) {
+		if (err == DIGEST_SINK_ERR_DIGEST_MISMATCH) {
+			LOG_ERR("Candidate inconsistent");
+		} else {
+			LOG_ERR("Failed to calculate digest: %d", err);
+		}
+
+		return SUIT_PLAT_ERR_CRASH;
+	}
+
+	LOG_DBG("Candidate consistent");
+
+	/* Then compare candidate's digest with current SDFW Recovery digest */
+	err = verify_digest(candidate_binary_start,
```

Comment:
Can't we use `memcmp` between `current_sdfw_recovery_digest` and `candidate_digest_in_manifest`  - as it is currently done in SDFW sink ? Comparing it to the whole digest seems a waste of time here.

Same for `check_recovery_activated`.

Reply:
I think it's better to keep this implementation of checking the digest and optimize later if the performance ever turns out to be a problem.

Reply:
Given our deadlines, I would prefer to leave current implementation as it is and optimize it in future a e-rk suggests.